python/pylibcudf/pylibcudf/column.pyx

```diff
@@ -1319,3 +1347,263 @@ def is_c_contiguous(
+    cdef size_t n = <size_t>arr.length
+    cdef list out = <list>PyList_New(<Py_ssize_t>n)
+    if out is None:
+        raise MemoryError("Unable to create new python List")
+
+    # buffers[0] = null-bitmap (optional)
+    # buffers[1] = int32 offsets
+    # buffers[2] = char* data
+    cdef const void** buffers = arr.buffers
+    cdef int64_t offset = arr.offset
+    cdef const uint8_t* null_bm = NULL
+    cdef const int32_t* offs = NULL
+    cdef const char* data = NULL
+    cdef size_t i, bit_index
+    cdef object s
+
+    if n == 0:
+        return out
+
+    if arr.n_buffers >= 1 and buffers[0] != NULL and arr.null_count != 0:
+        null_bm = <const uint8_t*>buffers[0]
+
+    if arr.n_buffers < 3 or buffers[1] == NULL or buffers[2] == NULL:
+        for i in range(n):
+            _set_item(out, <Py_ssize_t>i, None)
+        return out
+
+    offs = <const int32_t*>buffers[1]
+    data = <const char*>buffers[2]
+    offs = offs + <Py_ssize_t>offset
+
+    for i in range(n):
+        bit_index = <size_t>(offset + <int64_t>i)
+        if not _is_valid(null_bm, bit_index):
+            _set_item(out, <Py_ssize_t>i, None)
+            continue
+
+        s = PyUnicode_DecodeUTF8(data + offs[i],
+                                 <Py_ssize_t>(offs[i + 1] - offs[i]),
+                                 NULL)
+        if s is None:
+            raise MemoryError("Unable to create a python type from C type")
+        _set_item(out, <Py_ssize_t>i, s)
+
+    return out
+
+
+cdef list _arrow_to_pylist_numeric(type_id dtype, ArrowArray* arr):
+    cdef size_t n = <size_t>arr.length
+    cdef list out = <list>PyList_New(<Py_ssize_t>n)
+    if out is None:
+        raise MemoryError("Unable to create new python List")
+
+    cdef const void** buffers = arr.buffers
+    cdef int64_t offset = arr.offset
+    cdef const uint8_t* null_bm = NULL
+    cdef size_t i, bit_index
+    cdef size_t item_size
+    cdef object obj
+
+    if n == 0:
+        return out
+
+    if arr.n_buffers >= 1 and buffers[0] != NULL and arr.null_count != 0:
+        null_bm = <const uint8_t*>buffers[0]
+
+    if arr.n_buffers < 2 or buffers[1] == NULL:
+        for i in range(n):
+            _set_item(out, <Py_ssize_t>i, None)
+        return out
+
+    if dtype == type_id.INT8 or dtype == type_id.UINT8:
+        item_size = sizeof(uint8_t)
+    elif dtype == type_id.INT16  or dtype == type_id.UINT16:
+        item_size = sizeof(uint16_t)
+    elif dtype == type_id.INT32  or dtype == type_id.UINT32:
+        item_size = sizeof(uint32_t)
+    elif dtype == type_id.INT64  or dtype == type_id.UINT64:
+        item_size = sizeof(uint64_t)
+    elif dtype == type_id.FLOAT32:
+        item_size = sizeof(float)
+    elif dtype == type_id.FLOAT64:
+        item_size = sizeof(double)
+    else:
+        raise NotImplementedError(f"Column with {dtype=} not supported")
+
+    cdef const char* base = <const char*>buffers[1] + item_size * <size_t>offset
+
+    if null_bm is NULL:
```

Comment:
I see what you're trying to optimize here, but until we have a use case where we see the performance impact of checking the bitmask nullity N times instead of once I'm inclined to say duplicating the entire if/else cascade for types isn't worth it.